pkg/groupsync/groups.go

```diff
@@ -181,28 +182,47 @@ func (g *GroupMember) User() (*User, error) {
 // group ID using the given memberFunc. This function serves mostly as
 // a utility function when implementing ReadGroupClients for when there
 // is no special logic for fetching descendants.
-func Descendants(ctx context.Context, groupID string, memberFunc func(context.Context, string) ([]Member, error)) ([]*User, error) {
-	members, err := memberFunc(ctx, groupID)
-	if err != nil {
-		return nil, fmt.Errorf("error fetching group members: %s, %w", groupID, err)
-	}
-	var users []*User
+func Descendants(ctx context.Context, rootGroupID string, memberFunc func(context.Context, string) ([]Member, error)) ([]*User, error) {
+	// Need to do a BFS traversal of the group structure
+	queue := list.New()
```

Comment:
How about using a `[]string` rather than a list? That avoids the type assertion, and will be more familiar to readers.

Push would be `queue = append(queue, elem)` and pop would be `elem, queue = queue[0], queue[1:]`

Reply:
Couldn't that be potentially less efficient? I know Go can do `s = s[1:]` as just modification to view of the underlying array. But every time we append we may potentially have to allocate a new block of memory if the underlying array doesn't have any more slots.

Reply:
That's generally considered negligible. The Go runtime will grow the backing array in chunks so you're not constantly calling the allocator. Unless this is super performance critical (nanoseconds matter) or will be used on huge data sets, I think we should lean toward convention over optimization anyway. For example, I've looked at a lot of Go code and I didn't even realize there was a `container/list` until this PR :)

Reply:
ok, sounds good. I will use a slice then.